File: woodstorage/src/main/java/com/jordylangen/woodstorage/view/LogEntryAdapter.java
package com.jordylangen.woodstorage.view;

import android.annotation.SuppressLint;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.jordylangen.woodstorage.LogEntry;
import com.jordylangen.woodstorage.R;
import com.jordylangen.woodstorage.utils.ColorUtils;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public class LogEntryAdapter extends RecyclerView.Adapter<LogEntryAdapter.LogEntryViewHolder> {

    @SuppressLint("ConstantLocale")
    private static final SimpleDateFormat TIMESTAMP_FORMAT = new SimpleDateFormat("dd-MM-yyyy HH:mm:ss", Locale.getDefault());

    private List<LogEntry> logs = new ArrayList<>();
    private Map<String, Integer> tagColors = new HashMap<>();

    @Override
    public LogEntryViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_log, parent, false);
        return new LogEntryViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull LogEntryViewHolder holder, int position) {
        LogEntry log = logs.get(position);
        holder.tagTextView.setText(log.getTag());
        holder.messageTextView.setText(log.getMessage());
        holder.priorityTextView.setText(getPriorityTextResource(log));
        holder.timestampTextView.setText(TIMESTAMP_FORMAT.format(log.getTimeStamp()));

        int color;
        if (tagColors.containsKey(log.getTag())) {
            color = tagColors.get(log.getTag());
        } else {
            color = ColorUtils.randomColor();
            tagColors.put(log.getTag(), color);
        }

        holder.colorIndicatorView.setBackgroundColor(color);
    }

    private int getPriorityTextResource(LogEntry log) {
        switch (log.getPriority()) {
            case Log.DEBUG:
                return R.string.log_level_debug;
            case Log.INFO:
                return R.string.log_level_info;
            case Log.WARN:
                return R.string.log_level_warn;
            case Log.ERROR:
                return R.string.log_level_error;
            default:
                return R.string.log_level_verbose;
        }
    }

    @Override
    public int getItemCount() {
        return logs.size();
    }

    public void add(LogEntry logEntry) {
        logs.add(logEntry);
        notifyItemInserted(logs.size());
    }

    public void add(LogEntry logEntry, int index) {
        logs.add(index, logEntry);
        notifyDataSetChanged();
    }

    public void clear() {
        logs.clear();
        notifyDataSetChanged();
    }

    class LogEntryViewHolder extends RecyclerView.ViewHolder {

        private View colorIndicatorView;
        private TextView tagTextView;
        private TextView priorityTextView;
        private TextView messageTextView;
        private TextView timestampTextView;

        LogEntryViewHolder(View itemView) {
            super(itemView);

            colorIndicatorView = itemView.findViewById(R.id.log_color_indicator);
            tagTextView = itemView.findViewById(R.id.log_tag);
            priorityTextView = itemView.findViewById(R.id.log_priority);
            messageTextView = itemView.findViewById(R.id.log_message);
            timestampTextView = itemView.findViewById(R.id.log_timestamp);
        }
    }
}

File: woodstorage/src/main/java/com/jordylangen/woodstorage/view/LogEntryAdapter.kt
package com.jordylangen.woodstorage.view

import android.annotation.SuppressLint
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.jordylangen.woodstorage.storage.LogEntry
import com.jordylangen.woodstorage.R
import com.jordylangen.woodstorage.view.utils.ColorUtils

import java.text.SimpleDateFormat
import java.util.ArrayList
import java.util.HashMap
import java.util.Locale
import androidx.recyclerview.widget.RecyclerView

class LogEntryAdapter : RecyclerView.Adapter<LogEntryAdapter.LogEntryViewHolder>() {

    companion object {
        @SuppressLint("ConstantLocale")
        private val TIMESTAMP_FORMAT = SimpleDateFormat("dd-MM-yyyy HH:mm:ss", Locale.getDefault())
    }

    private val logs = ArrayList<LogEntry>()
    private val tagColors = HashMap<String, Int>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): LogEntryViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.view_log, parent, false)
        return LogEntryViewHolder(view)
    }

    override fun onBindViewHolder(holder: LogEntryViewHolder, position: Int) {
        val log = logs[position]
        holder.tagTextView.text = log.tag
        holder.messageTextView.text = log.message
        holder.priorityTextView.setText(getPriorityTextResource(log))
        holder.timestampTextView.text = TIMESTAMP_FORMAT.format(log.timeStamp)

        val color: Int
        if (tagColors.containsKey(log.tag)) {
            color = tagColors[log.tag]!!
        } else {
            color = ColorUtils.randomColor()
            tagColors[log.tag!!] = color
        }

        holder.colorIndicatorView.setBackgroundColor(color)
    }

    private fun getPriorityTextResource(log: LogEntry): Int {
        return when (log.priority) {
            Log.DEBUG -> R.string.log_level_debug
            Log.INFO -> R.string.log_level_info
            Log.WARN -> R.string.log_level_warn
            Log.ERROR -> R.string.log_level_error
            else -> R.string.log_level_verbose
        }
    }

    override fun getItemCount(): Int {
        return logs.size
    }

    fun add(logEntry: LogEntry) {
        logs.add(logEntry)
        notifyItemInserted(logs.size)
    }

    fun add(logEntry: LogEntry, index: Int) {
        logs.add(index, logEntry)
        notifyDataSetChanged()
    }

    fun clear() {
        logs.clear()
        notifyDataSetChanged()
    }

    inner class LogEntryViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val colorIndicatorView: View = itemView.findViewById(R.id.log_color_indicator)
        val tagTextView: TextView = itemView.findViewById(R.id.log_tag)
        val priorityTextView: TextView = itemView.findViewById(R.id.log_priority)
        val messageTextView: TextView = itemView.findViewById(R.id.log_message)
        val timestampTextView: TextView = itemView.findViewById(R.id.log_timestamp)
    }
}
